### PR TITLE
MSSG code via Github review

### PR DESCRIPTION
**Pull request by lauralorenz**
I've cloned the repository from [jeevan_shankar's bitbucket repo](https://bitbucket.org/jeevan_shankar/multi-sense-skipgram/overview) with the code related to his paper ["Efficient Non-parametric Estimation of Multiple Embeddings per Word in Vector Space"](https://arxiv.org/pdf/1504.06654v1.pdf) that our RL is reading for discussion this week. This PR is an artificial way to zero in on the source code files from that repository most relevant to review for our reading of his paper, and gives us the opportunity to utilize Github's line commenting/code review features to have a discussion against the code without having to change the source at all.

### CODE REVIEW

File: make_cp.sh [hunk 1]
```diff
@@ -1,9 +0,0 @@
-#!/bin/sh
```
Comment (lauralorenz) [inline make_cp.sh:1]: Way to format the path for the scripts that will call the compiled scala

File: src/HogWildTrainer.scala [hunk 2]
```diff
@@ -1,22 +0,0 @@
-
-import cc.factorie.optimize.{ GradientOptimizer, Trainer, Example }
-import cc.factorie.la.SmartGradientAccumulator
-import cc.factorie.util.{ LocalDoubleAccumulator, Threading }
-import cc.factorie.model.WeightsSet
-
-class HogWildTrainer(val weightsSet: WeightsSet, val optimizer: GradientOptimizer, val nThreads: Int = Runtime.getRuntime.availableProcessors(), val maxIterations: Int = 3)
-  extends Trainer {
```
Comment (lauralorenz) [inline src/HogWildTrainer.scala:8]: `HogWildTrainer` extends from the `Trainer` class defined in `factorie`, a Scala neural network library: https://github.com/factorie/factorie/blob/master/src/main/scala/cc/factorie/optimize/Trainer.scala#L382

File: src/MultiSenseSkipGramEmbeddingModel.scala [hunk 3]
```diff
@@ -1,190 +0,0 @@
...
-  def subSample(word: Int): Int = {
-    val ran = vocab.getSubSampleProb(word) // see the vocabBuilder to understand how this sub-sample prob is got
-    val real_ran = rng.nextInt(0xFFFF) / 0xFFFF.toDouble
-    return if (ran < real_ran) -1 else word
-  }
-}
-
-
-
-class MSCBOWSkipGramNegSamplingExample(model: MultiSenseWordEmbeddingModel, word: Int, sense : Int, context : Int, label: Int) extends Example {
```
Comment (lauralorenz) [inline src/MultiSenseSkipGramEmbeddingModel.scala:154]: Feed forward for MSSG `Example`

File: src/HogWildTrainer.scala [hunk 4]
```diff
@@ -1,22 +0,0 @@
...
-
-  var iteration = 0
-  def processExample(e: Example): Unit = {
```
Comment (lauralorenz) [inline src/HogWildTrainer.scala:11]: feed forward against an `Example` with the `Example`'s `accumulateValueAndGradient` method and allow the optimizer to step the gradient

File: src/SkipGramEmbeddingModel.scala [hunk 5]
```diff
@@ -1,75 +0,0 @@
...
-    val ran = vocab.getSubSampleProb(word) // pre-computed to avoid sqrt call every time. Improvement of 10 secs on 100MB data ~ 15 MINs on 10GB
-    //val cnt = vocab.getCount(word)
-    //val ran = (math.sqrt(cnt / (sample * train_words)) + 1) * (sample * train_words) / cnt
-    val real_ran = rng.nextInt(0xFFFF) / 0xFFFF.toDouble
-    if (ran < real_ran) { return -1 }
-    else return word
-  }
-}
-
-class SkipGramNegSamplingExample(model: WordEmbeddingModel, word: Int, context: Int, label: Int) extends Example {
```
Comment (lauralorenz) [inline src/SkipGramEmbeddingModel.scala:50]: Feed forward for vanilla `Skip-Gram` (with neg sampling) with `Example`

File: src/MultiSenseWordEmbeddingModel.scala [hunk 6]
```diff
@@ -1,239 +0,0 @@
...
-import cc.factorie.util.Threading
-import java.io.{File, PrintWriter}
-import java.util.zip.GZIPInputStream
-import java.io.FileInputStream
-import java.io.OutputStreamWriter
-import java.util.zip.GZIPOutputStream
-import java.io.BufferedOutputStream
-import java.io.FileOutputStream
-
-abstract class MultiSenseWordEmbeddingModel(val opts: EmbeddingOpts) extends Parameters {
```
Comment (lauralorenz) [inline src/MultiSenseWordEmbeddingModel.scala:13]: Interface for all multi sense models (e.g. MSSG, NPMSSG). They share vocab building, model persistence. Seems to me that the shared `learnEmbeddings` method is a wrapper for a model's threaded `processExamples` implementation.

File: src/MultiSenseWordEmbeddingModel.scala [hunk 7]
```diff
@@ -1,239 +0,0 @@
...
-  protected var clusterCenter: Array[Array[DenseTensor1]] = null   
-  protected var clusterCount: Array[Array[Int]]           = null 
-  protected var clusterActive: Array[Array[Int]] = null
-  protected var ncluster: Array[Int] = null // holds the information for # of cluster
-  protected var countSenses: Array[Int] = null
-  protected var learnMultiVec: Array[Boolean] = null
-
-  // Component-1
-  def buildVocab(minFreq: Int = 5): Unit = {
-    vocab = new VocabBuilder(vocabHashSize, samplingTableSize, 0.7) // 0.7 is the load factor
```
Comment (lauralorenz) [inline src/MultiSenseWordEmbeddingModel.scala:65]: vocab actually built with this other class `VocabBuilder`

File: src/VocabBuilder.scala [hunk 8]
```diff
@@ -1,227 +0,0 @@
...
-  private val power: Double = 0.75
-
-  // assign internal params 
-  private val rng = new Random(5) // fix the seed
-  private var train_words = 0
-  private var prev_vocab_size = 0
-
-  // allocate data structures
-  private var vocab_hash = new Array[Int](vocab_hash_size) // vocab_hash stores pointer to the vocab array 
-  private var vocab = new Array[vocab_word](vocab_max_size) //  vocab array holds the (word, cnt) pairs. In future, huffman encoding will be added.
```
Comment (lauralorenz) [inline src/VocabBuilder.scala:26]: All my reading suggests these are stored in this format but I'm not sure how/if they get converted to one hot before feed forward and I'm also curious at what point word sense vectors aka v(s)(w(t),k) get converted to one hot as well (see Figure 2).

File: src/preprocess.scala [hunk 9]
```diff
@@ -1,35 +0,0 @@
-import cc.factorie.app.nlp.segment.DeterministicTokenizer
-import cc.factorie.app.nlp.segment.DeterministicSentenceSegmenter
-import cc.factorie.app.nlp.Document
-import java.io.PrintWriter
-import java.nio.charset.Charset
-
-object preprocess {
```
Comment (lauralorenz) [inline src/preprocess.scala:7]: NLP preprocessing like stemming against the source corpus.

File: train-mssg.sh [hunk 10]
```diff
@@ -1,16 +0,0 @@
-#!/bin/sh
-
-if [ ! -e text8 ]; then
-  wget https://dl.dropboxusercontent.com/u/39534006/text8.zip
```
Comment (lauralorenz) [inline train-mssg.sh:4]: wrangled corpus derived from Wikipedia English often used for text compression or processing benchmarks: [see details here](http://www.mattmahoney.net/dc/textdata)

File: src/EmbeddingOpts.scala [hunk 11]
```diff
@@ -1,56 +0,0 @@
-//package cc.factorie.app.nlp.embeddings
-import cc.factorie.util.CmdOptions
-
-class EmbeddingOpts extends CmdOptions {
```
Comment (lauralorenz) [inline src/EmbeddingOpts.scala:4]: Opts passed to each model class to configure things such as optimizer learning rate, negative sampling rate, etc.

File: src/HogWildTrainer.scala [hunk 12]
```diff
@@ -1,22 +0,0 @@
-
-import cc.factorie.optimize.{ GradientOptimizer, Trainer, Example }
-import cc.factorie.la.SmartGradientAccumulator
-import cc.factorie.util.{ LocalDoubleAccumulator, Threading }
-import cc.factorie.model.WeightsSet
-
-class HogWildTrainer(val weightsSet: WeightsSet, val optimizer: GradientOptimizer, val nThreads: Int = Runtime.getRuntime.availableProcessors(), val maxIterations: Int = 3)
```
Comment (lauralorenz) [inline src/HogWildTrainer.scala:7]: HogWild seems to be a design paradigm for SGD and related optimizers that improves computing efficiency. [See this paper](https://people.eecs.berkeley.edu/~brecht/papers/hogwildTR.pdf).

File: src/HogWildTrainer.scala (hunk 4 above)
Comment (lauralorenz) [reply src/HogWildTrainer.scala:11]: Each `Example` knows how to feed forward itself and apply its own gradients, in the way that is unique to its model implementation

File: src/WordEmbeddingModel.scala [hunk 13]
```diff
@@ -1,130 +0,0 @@
-//package cc.factorie.app.nlp.embeddings
-import cc.factorie.model.{Parameters, Weights}
-import cc.factorie.optimize.{Trainer, AdaGradRDA}
-import cc.factorie.la.DenseTensor1
-import cc.factorie.util.Threading
-import java.io.{File, PrintWriter, OutputStreamWriter, FileOutputStream, FileInputStream, BufferedOutputStream}
-import java.util.zip.{GZIPOutputStream, GZIPInputStream}
-
-
-abstract class WordEmbeddingModel(val opts: EmbeddingOpts) extends Parameters {
```
Comment (lauralorenz) [inline src/WordEmbeddingModel.scala:10]: Interface for Skip-Gram based models (in here there is a vanilla SkipGram with neg sampling). Also contains persistence storage method, vocab building method, and `learnEmbeddings` method which seems to be a wrapper for a model's threaded `processExamples` implementation.

File: src/WordVec.scala [hunk 14]
```diff
@@ -1,31 +0,0 @@
-// package cc.factorie.app.nlp.embeddings
-import java.nio.charset.Charset
-
-object WordVec {
```
Comment (lauralorenz) [inline src/WordVec.scala:4]: Entrypoint for scripts to pick a model and generate word embeddings.
